primitives/consensus/beefy/src/lib.rs

```diff
@@ -302,6 +337,56 @@ where
 	return valid_first && valid_second
 }
 
+/// Validates [ForkEquivocationProof] by checking:
+/// 1. `commitment` is signed,
+/// 2. `correct_header` is valid and matches `commitment.block_number`.
+/// 2. `commitment.payload` != `expected_payload(correct_header)`.
+/// NOTE: GRANDPA finalization proof is not checked, which leads to slashing on forks.
+/// This is fine since honest validators will not be slashed on the chain finalized
+/// by GRANDPA, which is the only chain that ultimately matters.
+/// The only material difference not checking GRANDPA proofs makes is that validators
+/// are not slashed for signing BEEFY commitments prior to the blocks committed to being
+/// finalized by GRANDPA. This is fine too, since the slashing risk of committing to
+/// an incorrect block implies validators will only sign blocks they *know* will be
+/// finalized by GRANDPA.
+pub fn check_fork_equivocation_proof<Number, Id, MsgHash, Header>(
```

Comment:
this is ok for now (while only checking header hashes) to live in primitives, but when moving to MMR root membership proofs this should be some `pallet_beefy::Config` trait type that is provided by `pallet_beefy_mmr` (or some other generic, pluggable solution).